onnxruntime/core/providers/cuda/math/unary_elementwise_ops_impl.cu

```diff
@@ -115,50 +119,139 @@ struct OP_Cast {
+IMPL_CAST_IMPL_FROM(uint32_t)
+IMPL_CAST_IMPL_FROM(uint64_t)
+IMPL_CAST_IMPL_FROM(bool)
+IMPL_CAST_IMPL_FROM(BFloat16)
+IMPL_CAST_IMPL_FROM(Float8E4M3FN)
+IMPL_CAST_IMPL_FROM(Float8E5M2)
+
+template <typename InT, typename OutT>
+struct OP_CastSat {
+  __device__ __inline__ OutT operator()(const InT& a) const;
+};
+
 template <typename InT, typename OutT>
-void Impl_Cast(
-    cudaStream_t stream,
-    const InT* input_data,
-    OutT* output_data,
-    size_t count) {
-  UnaryElementWiseImpl(stream,
-                       input_data,
-                       output_data,
-                       OP_Cast<InT, OutT>(),
-                       count);
-}
-
-#define SPECIALIZED_CAST_IMPL2(InT, OutT) \
-  template void Impl_Cast<InT, OutT>(cudaStream_t stream, const InT* input_data, OutT* output_data, size_t count);
-
-#define SPECIALIZED_CAST_FROM(T)      \
-  SPECIALIZED_CAST_IMPL2(T, half)     \
-  SPECIALIZED_CAST_IMPL2(T, float)    \
-  SPECIALIZED_CAST_IMPL2(T, double)   \
-  SPECIALIZED_CAST_IMPL2(T, int8_t)   \
-  SPECIALIZED_CAST_IMPL2(T, int16_t)  \
-  SPECIALIZED_CAST_IMPL2(T, int32_t)  \
-  SPECIALIZED_CAST_IMPL2(T, int64_t)  \
-  SPECIALIZED_CAST_IMPL2(T, uint8_t)  \
-  SPECIALIZED_CAST_IMPL2(T, uint16_t) \
-  SPECIALIZED_CAST_IMPL2(T, uint32_t) \
-  SPECIALIZED_CAST_IMPL2(T, uint64_t) \
-  SPECIALIZED_CAST_IMPL2(T, bool)     \
-  SPECIALIZED_CAST_IMPL2(T, BFloat16)
-
-SPECIALIZED_CAST_FROM(half)
-SPECIALIZED_CAST_FROM(float)
-SPECIALIZED_CAST_FROM(double)
-SPECIALIZED_CAST_FROM(int8_t)
-SPECIALIZED_CAST_FROM(int16_t)
-SPECIALIZED_CAST_FROM(int32_t)
-SPECIALIZED_CAST_FROM(int64_t)
-SPECIALIZED_CAST_FROM(uint8_t)
-SPECIALIZED_CAST_FROM(uint16_t)
-SPECIALIZED_CAST_FROM(uint32_t)
-SPECIALIZED_CAST_FROM(uint64_t)
-SPECIALIZED_CAST_FROM(bool)
-SPECIALIZED_CAST_FROM(BFloat16)
+struct OP_CastNoSat {
+  __device__ __inline__ OutT operator()(const InT& a) const;
+};
+
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 11080
+
+#define OP_CAST(T, NVT) \
+  template <> \
+  struct OP_CastSat<half, T> { \
+    __device__ __inline__ T operator()(const half& v) const { \
+      return T(static_cast<unsigned char>(__nv_cvt_halfraw_to_fp8(v, __NV_SATFINITE, NVT)), T::FromBits()); \
+    } \
+  }; \
+  template <> \
+  struct OP_CastNoSat<half, T> { \
+    __device__ __inline__ T operator()(const half& v) const { \
+      return T(static_cast<unsigned char>(__nv_cvt_halfraw_to_fp8(v, __NV_NOSAT, NVT)), T::FromBits()); \
+    } \
+  }; \
+  template <> \
+  struct OP_CastSat<float, T> { \
+    __device__ __inline__ T operator()(const float& v) const { \
+      return T(static_cast<unsigned char>(__nv_cvt_float_to_fp8(v, __NV_SATFINITE, NVT)), T::FromBits()); \
+    } \
+  }; \
+  template <> \
+  struct OP_CastNoSat<float, T> { \
+    __device__ __inline__ T operator()(const float& v) const { \
+      return T(static_cast<unsigned char>(__nv_cvt_float_to_fp8(v, __NV_NOSAT, NVT)), T::FromBits()); \
+    } \
+  };
+
+#else
+
+#define OP_CAST(T, NVT) \
```

Comment:
Have you tested this locally? All our build vms now have cuda 11.8 so this path will not be executed.

Reply:
Yes but not recently. I try again. one of dev machine has still cuda 11.6.